nominal/core/multiprocessed_stream.py

```diff
@@ -0,0 +1,190 @@
+from __future__ import annotations
+
+import concurrent.futures
+import logging
+import threading
+from concurrent.futures import ProcessPoolExecutor, ThreadPoolExecutor
+from dataclasses import dataclass, field
+from datetime import datetime, timedelta
+from queue import Queue
+from types import TracebackType
+from typing import Callable, Sequence, Type
+
+from typing_extensions import Self
+
+from nominal.core._clientsbunch import ProtoWriteService
+from nominal.core.queueing import (
+    QueueShutdown,
+    ReadQueue,
+    iter_queue,
+    spawn_batching_thread,
+)
+from nominal.core.stream import BatchItem
+from nominal.ts import IntegralNanosecondsUTC
+
+logger = logging.getLogger(__name__)
+
+
+@dataclass()
+class MultiProcessedWriteStream:
+    _serialize_batch: Callable[[Sequence[BatchItem]], bytes]
+    _max_batch_size: int
+    _max_wait: timedelta
+    _max_queue_size: int
+    _item_queue: Queue[BatchItem | QueueShutdown]
+    _batch_queue: ReadQueue[Sequence[BatchItem]]
+    _batch_thread: threading.Thread
+    _process_pool: ProcessPoolExecutor
+    _thread_pool: ThreadPoolExecutor
+    _client_factory: Callable[[], ProtoWriteService]
+    _auth_header: str
+    _nominal_data_source_rid: str
+    _process_thread: threading.Thread | None = field(default=None)
+
+    @classmethod
+    def create(
+        cls,
+        nominal_data_source_rid: str,
+        serialize_batch: Callable[[Sequence[BatchItem]], bytes],
+        max_batch_size: int,
+        max_wait: timedelta,
+        max_queue_size: int,
+        client_factory: Callable[[], ProtoWriteService],
+        auth_header: str,
+        max_workers: int,
+    ) -> Self:
+        """Create a new MultiProcessedWriteStream instance.
+
+        Args:
+            nominal_data_source_rid: Nominal data source rid
+            serialize_batch: Function to serialize batches of items
+            max_batch_size: How many items to accumulate before serializing
+            max_wait: Maximum time to wait before serializing a partial batch
+            max_queue_size: Maximum number of items that can be queued (0 for unlimited)
+            client_factory: Factory function to create ProtoWriteService instances
+            auth_header: Authentication header
+            max_workers: Maximum number of worker threads for parallel processing
+        """
+        process_pool = ProcessPoolExecutor(max_workers=max_workers)
+        thread_pool = ThreadPoolExecutor()
+        item_maxsize = max_queue_size if max_queue_size > 0 else 0
+        batch_maxsize = (max_queue_size // max_batch_size) if max_queue_size > 0 else 0
+
+        item_queue: Queue[BatchItem | QueueShutdown] = Queue(maxsize=item_maxsize)
+
+        batch_thread, batch_queue = spawn_batching_thread(
+            item_queue,
+            max_batch_size,
+            max_wait,
+            max_queue_size=batch_maxsize,
+        )
+
+        instance = cls(
+            _serialize_batch=serialize_batch,
```

Comment:
It'd be nice if we turned this into something like a `Serializer` type that could have a multiprocessing, multithreading, or main thread implementation